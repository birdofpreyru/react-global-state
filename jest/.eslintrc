{
  "rules": {
    "import/no-extraneous-dependencies": ["error", {}],
    "no-console": "off"
  }
}
