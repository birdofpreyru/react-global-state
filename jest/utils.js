/**
 * Jest test utils.
 */

import _ from 'lodash';
import mockdate from 'mockdate';
import { render, unmountComponentAtNode } from 'react-dom';

export { act } from 'react-dom/test-utils';

/**
 * Generates a mock UUID.
 * @param {Number} seed
 * @return {String}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

const originalConsole = console.log;

export async function mockConsoleLog() {
  console.log = (...args) => console.log.logs.push(_.cloneDeep(args));
  console.log.clear = () => {
    console.log.logs = [];
  };
  console.log.clear();
}

export async function unMockConsoleLog() {
  console.log = originalConsole;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {Number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param {React.ReactNode} scene
 * @return {HTMLElement}
 */
export function mount(scene) {
  const res = document.createElement('div');
  document.body.appendChild(res);
  render(scene, res);
  return res;
}

/**
 * Waits for `time`.
 * @param {Number} time Time [ms].
 * @returns {Promise} Resolves once `time` have passed.
 */
export async function timer(time) {
  return new Promise((resolve) => {
    setTimeout(resolve, time);
  });
}

/**
 * Unmounts `scene` from the DOM.
 * @param {HTMLElement} scene
 */
export function unmount(scene) {
  unmountComponentAtNode(scene);
  scene.remove();
}
