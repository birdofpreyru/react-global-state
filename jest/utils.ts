/**
 * Jest test utils.
 */

import { cloneDeep } from 'lodash';
import mockdate from 'mockdate';
import { type ReactNode } from 'react';
import { type Root, createRoot } from 'react-dom/client';
import { act } from 'react-dom/test-utils';

export { act };

(global as { [key: string]: unknown }).IS_REACT_ACT_ENVIRONMENT = true;

/**
 * Generates a mock UUID.
 * @param seed
 * @return
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

export const consoleLogs: unknown[] = [];

const originalConsole = console.log;

export async function mockConsoleLog() {
  console.log = (...args) => consoleLogs.push(cloneDeep(args));
  console.groupCollapsed = console.log;
  consoleLogs.splice(0);
}

export async function unMockConsoleLog() {
  console.log = originalConsole;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param time Time step [ms].
 * @returns Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time: number) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

export type DestroyableHtmlElement = HTMLElement & { destroy: () => void };

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param scene
 * @return {HTMLElement} Created container DOM element with unmount
 *  function attached.
 */
export function mount(scene: ReactNode): DestroyableHtmlElement {
  let root: Root;

  const rootElement = document.createElement('div');

  const res: DestroyableHtmlElement = (rootElement as unknown) as DestroyableHtmlElement;

  res.destroy = () => {
    act(() => root.unmount());
    res.remove();
  };

  document.body.appendChild(rootElement);
  act(() => {
    root = createRoot(rootElement);
    root.render(scene);
  });
  return res;
}

/**
 * Advances mock timers by the specified time, and waits for any DOM updates
 * to complete.
 * @param time [miliseconds]
 * @return Wait for this.
 */
export function wait(time: number) {
  return act(() => mockTimer(time));
}
