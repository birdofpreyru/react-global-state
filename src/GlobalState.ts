import {
  get,
  isFunction,
  isObject,
  isNil,
  set,
  toPath,
} from 'lodash';

import SsrContext from './SsrContext';

import {
  type CallbackT,
  type ForceT,
  type LockT,
  type TypeLock,
  type ValueAtPathT,
  type ValueOrInitializerT,
  cloneDeepForLog,
  isDebugMode,
} from './utils';

const ERR_NO_SSR_WATCH = 'GlobalState must not be watched at server side';

type GetOptsT<T> = {
  initialState?: boolean;
  initialValue?: ValueOrInitializerT<T>;
};

export default class GlobalState<
  StateT,
  SsrContextT extends SsrContext<StateT> = SsrContext<StateT>,
> {
  readonly ssrContext?: SsrContextT;

  #dependencies: { [key: string]: Readonly<any[]> } = {};

  #initialState: StateT;

  // TODO: It is tempting to replace watchers here by
  // Emitter from @dr.pogodin/js-utils, but we need to clone
  // current watchers for emitting later, and this is not something
  // Emitter supports right now.
  #watchers: CallbackT[] = [];

  #nextNotifierId?: NodeJS.Timeout;

  #currentState: StateT;

  /**
   * Creates a new global state object.
   * @param initialState Intial global state content.
   * @param ssrContext Server-side rendering context.
   */
  constructor(
    initialState: StateT,
    ssrContext?: SsrContextT,
  ) {
    this.#currentState = initialState;
    this.#initialState = initialState;

    if (ssrContext) {
      /* eslint-disable no-param-reassign */
      ssrContext.dirty = false;
      ssrContext.pending = [];
      ssrContext.state = this.#currentState;
      /* eslint-enable no-param-reassign */

      this.ssrContext = ssrContext;
    }

    if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
      /* eslint-disable no-console */
      let msg = 'New ReactGlobalState created';
      if (ssrContext) msg += ' (SSR mode)';
      console.groupCollapsed(msg);
      console.log('Initial state:', cloneDeepForLog(initialState));
      console.groupEnd();
      /* eslint-enable no-console */
    }
  }

  /**
   * Drops the record of dependencies, if any, for the given path.
   */
  dropDependencies(path: string) {
    delete this.#dependencies[path];
  }

  /**
   * Checks if given `deps` are different from previously recorded ones for
   * the given `path`. If they are, `deps` are recorded as the new deps for
   * the `path`, and also the array is frozen, to prevent it from being
   * modified.
   */
  hasChangedDependencies(path: string, deps: any[]): boolean {
    const prevDeps = this.#dependencies[path];
    let changed = !prevDeps || prevDeps.length !== deps.length;
    for (let i = 0; !changed && i < deps.length; ++i) {
      changed = prevDeps![i] !== deps[i];
    }
    this.#dependencies[path] = Object.freeze(deps);
    return changed;
  }

  /**
   * Gets entire state, the same way as .get(null, opts) would do.
   * @param opts.initialState
   * @param opts.initialValue
   */
  getEntireState(opts?: GetOptsT<StateT>): StateT {
    let state = opts?.initialState ? this.#initialState : this.#currentState;
    if (state !== undefined || opts?.initialValue === undefined) return state;

    const iv = opts.initialValue;
    state = isFunction(iv) ? iv() : iv;
    if (this.#currentState === undefined) this.setEntireState(state);
    return state;
  }

  /**
   * Notifies all connected state watchers that a state update has happened.
   */
  private notifyStateUpdate(path: null | string | undefined, value: unknown) {
    if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
      /* eslint-disable no-console */
      const p = typeof path === 'string'
        ? `"${path}"` : 'none (entire state update)';
      console.groupCollapsed(`ReactGlobalState update. Path: ${p}`);
      console.log('New value:', cloneDeepForLog(value, path ?? ''));
      console.log('New state:', cloneDeepForLog(this.#currentState));
      console.groupEnd();
      /* eslint-enable no-console */
    }

    if (this.ssrContext) {
      this.ssrContext.dirty = true;
      this.ssrContext.state = this.#currentState;
    } else if (!this.#nextNotifierId) {
      this.#nextNotifierId = setTimeout(() => {
        this.#nextNotifierId = undefined;
        const watchers = [...this.#watchers];
        for (let i = 0; i < watchers.length; ++i) {
          watchers[i]!();
        }
      });
    }
  }

  /**
   * Sets entire state, the same way as .set(null, value) would do.
   * @param value
   */
  setEntireState(value: StateT): StateT {
    if (this.#currentState !== value) {
      this.#currentState = value;
      this.notifyStateUpdate(null, value);
    }
    return value;
  }

  /**
   * Gets current or initial value at the specified "path" of the global state.
   * @param path Dot-delimitered state path.
   * @param options Additional options.
   * @param options.initialState If "true" the value will be read
   *  from the initial state instead of the current one.
   * @param options.initialValue If the value read from the "path" is
   *  "undefined", this "initialValue" will be returned instead. In such case
   *  "initialValue" will also be written to the "path" of the current global
   *  state (no matter "initialState" flag), if "undefined" is stored there.
   * @return Retrieved value.
   */

  // .get() without arguments just falls back to .getEntireState().
  get(): StateT;

  // This variant attempts to automatically resolve and check the type of value
  // at the given path, as precise as the actual state and path types permit.
  // If the automatic path resolution is not possible, the ValueT fallsback
  // to `never` (or to `undefined` in some cases), effectively forbidding
  // to use this .get() variant.
  get<
    PathT extends null | string | undefined,
    ValueArgT extends ValueAtPathT<StateT, PathT, never>,
    ValueResT extends ValueAtPathT<StateT, PathT, void>,
  >(path: PathT, opts?: GetOptsT<ValueArgT>): ValueResT;

  // This variant is not callable by default (without generic arguments),
  // otherwise it allows to set the correct ValueT directly.
  get<Forced extends ForceT | LockT = LockT, ValueT = void>(
    path?: null | string,
    opts?: GetOptsT<TypeLock<Forced, never, ValueT>>,
  ): TypeLock<Forced, void, ValueT>;

  get<ValueT>(path?: null | string, opts?: GetOptsT<ValueT>): ValueT {
    if (isNil(path)) {
      const res = this.getEntireState((opts as unknown) as GetOptsT<StateT>);
      return (res as unknown) as ValueT;
    }

    const state = opts?.initialState ? this.#initialState : this.#currentState;

    let res = get(state, path);
    if (res !== undefined || opts?.initialValue === undefined) return res;

    const iv = opts.initialValue;
    res = isFunction(iv) ? iv() : iv;

    if (!opts?.initialState || this.get(path) === undefined) {
      this.set<ForceT, unknown>(path, res);
    }

    return res;
  }

  /**
   * Writes the `value` to given global state `path`.
   * @param path Dot-delimitered state path. If not given, entire
   * global state content is replaced by the `value`.
   * @param value The value.
   * @return Given `value` itself.
   */

  // This variant attempts automatic value type resolution & checking.
  set<
    PathT extends null | string | undefined,
    ValueArgT extends ValueAtPathT<StateT, PathT, never>,
    ValueResT extends ValueAtPathT<StateT, PathT, void>,
  >(path: PathT, value: ValueArgT): ValueResT;

  // This variant is disabled by default, otherwise allows to give
  // expected value type explicitly.
  set<Forced extends ForceT | LockT = LockT, ValueT = never>(
    path: null | string | undefined,
    value: TypeLock<Forced, never, ValueT>,
  ): TypeLock<Forced, void, ValueT>;

  set(path: null | string | undefined, value: unknown): unknown {
    if (isNil(path)) return this.setEntireState(value as StateT);

    if (value !== this.get(path)) {
      const root = { state: this.#currentState };
      let segIdx = 0;
      let pos: any = root;
      const pathSegments = toPath(`state.${path}`);
      for (; segIdx < pathSegments.length - 1; segIdx += 1) {
        const seg = pathSegments[segIdx]!;
        const next = pos[seg];
        if (Array.isArray(next)) pos[seg] = [...next];
        else if (isObject(next)) pos[seg] = { ...next };
        else {
          // We arrived to a state sub-segment, where the remaining part of
          // the update path does not exist yet. We rely on lodash's set()
          // function to create the remaining path, and set the value.
          set(pos, pathSegments.slice(segIdx), value);
          break;
        }
        pos = pos[seg];
      }

      if (segIdx === pathSegments.length - 1) {
        pos[pathSegments[segIdx]!] = value;
      }

      this.#currentState = root.state;

      this.notifyStateUpdate(path, value);
    }
    return value;
  }

  /**
   * Unsubscribes `callback` from watching state updates; no operation if
   * `callback` is not subscribed to the state updates.
   * @param callback
   * @throws if {@link SsrContext} is attached to the state instance: the state
   * watching functionality is intended for client-side (non-SSR) only.
   */
  unWatch(callback: CallbackT) {
    if (this.ssrContext) throw new Error(ERR_NO_SSR_WATCH);

    const watchers = this.#watchers;
    const pos = watchers.indexOf(callback);
    if (pos >= 0) {
      watchers[pos] = watchers[watchers.length - 1]!;
      watchers.pop();
    }
  }

  /**
   * Subscribes `callback` to watch state updates; no operation if
   * `callback` is already subscribed to this state instance.
   * @param callback It will be called without any arguments every
   * time the state content changes (note, howhever, separate state updates can
   * be applied to the state at once, and watching callbacks will be called once
   * after such bulk update).
   * @throws if {@link SsrContext} is attached to the state instance: the state
   * watching functionality is intended for client-side (non-SSR) only.
   */
  watch(callback: CallbackT) {
    if (this.ssrContext) throw new Error(ERR_NO_SSR_WATCH);

    const watchers = this.#watchers;
    if (watchers.indexOf(callback) < 0) {
      watchers.push(callback);
    }
  }
}
