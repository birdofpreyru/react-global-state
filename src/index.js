export { default as GlobalStateProvider } from './GlobalStateProvider';
export { default as useAsyncData } from './useAsyncData';
export { default as useGlobalState } from './useGlobalState';
