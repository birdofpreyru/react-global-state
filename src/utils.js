// Auxiliary stuff.

/**
 * Returns 'true' if debug logging should be performed; 'false' otherwise.
 *
 * BEWARE: The actual safeguards for the debug logging still should read
 *  if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
 *    // Some debug logging
 *  }
 * to ensure that debug code is stripped out by Webpack in production mode.
 *
 * @returns {boolean}
 * @ignore
 */
export function isDebugMode() {
  try {
    return process.env.NODE_ENV !== 'production'
      && !!process.env.REACT_GLOBAL_STATE_DEBUG;
  } catch (error) {
    return false;
  }
}

export default null;
