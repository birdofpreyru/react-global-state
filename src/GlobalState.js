import {
  cloneDeep,
  get,
  isObject,
  isNil,
  set,
  toPath,
} from 'lodash';

import { isDebugMode } from './utils';

const ERR_NO_SSR_WATCH = 'GlobalState must not be watched at server side';

export default class GlobalState {
  #nextNotifierId = null;

  #state;

  #watchers = [];

  /**
   * Creates a new global state object.
   * @param {any} [initialState] Intial global state content.
   * @param {SsrContext} [ssrContext] Server-side rendering context.
   */
  constructor(initialState, ssrContext) {
    /* eslint-disable no-param-reassign */
    this.#state = initialState;

    if (ssrContext) {
      ssrContext.dirty = false;
      ssrContext.pending = [];
      ssrContext.state = this.#state;
      this.ssrContext = ssrContext;
    }

    if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
      /* eslint-disable no-console */
      let msg = 'New ReactGlobalState created';
      if (ssrContext) msg += ' (SSR mode)';
      console.groupCollapsed(msg);
      console.log('Initial state:', cloneDeep(initialState));
      console.groupEnd();
      /* eslint-enable no-console */
    }
    /* eslint-enable no-param-reassign */
  }

  /**
   * Gets the value at given `path` of global state. If `path` is null or
   * undefined, the entire state object is returned.
   * @param {string} [path] Dot-delimitered state path. If not given, entire
   * global state content is returned.
   * @return {any}
   */
  get(path) {
    return isNil(path) ? this.#state : get(this.#state, path);
  }

  /**
   * Writes the `value` to given global state `path`.
   * @param {string} [path] Dot-delimitered state path. If not given, entire
   * global state content is replaced by the `value`.
   * @param {any} value The value.
   * @return {any} Given `value` itself.
   */
  set(path, value) {
    if (value !== this.get(path)) {
      if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
        /* eslint-disable no-console */
        console.groupCollapsed(
          `ReactGlobalState update. Path: "${path || ''}"`,
        );
        console.log('New value:', cloneDeep(value));
        /* eslint-enable no-console */
      }

      if (isNil(path)) this.#state = value;
      else {
        const root = { state: this.#state };
        let segIdx = 0;
        let pos = root;
        const pathSegments = toPath(`state.${path}`);
        for (; segIdx < pathSegments.length - 1; segIdx += 1) {
          const seg = pathSegments[segIdx];
          const next = pos[seg];
          if (Array.isArray(next)) pos[seg] = [...next];
          else if (isObject(next)) pos[seg] = { ...next };
          else {
            // We arrived to a state sub-segment, where the remaining part of
            // the update path does not exist yet. We rely on lodash's set()
            // function to create the remaining path, and set the value.
            set(pos, pathSegments.slice(segIdx), value);
            break;
          }
          pos = pos[seg];
        }

        if (segIdx === pathSegments.length - 1) {
          pos[pathSegments[segIdx]] = value;
        }

        this.#state = root.state;
      }

      if (this.ssrContext) {
        this.ssrContext.dirty = true;
        this.ssrContext.state = this.#state;
      } else if (!this.#nextNotifierId) {
        this.#nextNotifierId = setTimeout(() => {
          this.#nextNotifierId = null;
          [...this.#watchers].forEach((w) => w());
        });
      }
      if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
        /* eslint-disable no-console */
        console.log('New state:', cloneDeep(this.#state));
        console.groupEnd();
        /* eslint-enable no-console */
      }
    }
    return value;
  }

  /**
   * Unsubscribes `callback` from watching state updates; no operation if
   * `callback` is not subscribed to the state updates.
   * @param {function} callback
   * @throws if {@link SsrContext} is attached to the state instance: the state
   * watching functionality is intended for client-side (non-SSR) only.
   */
  unWatch(callback) {
    if (this.ssrContext) throw new Error(ERR_NO_SSR_WATCH);

    const watchers = this.#watchers;
    const pos = watchers.indexOf(callback);
    if (pos >= 0) {
      watchers[pos] = watchers[watchers.length - 1];
      watchers.pop();
    }
  }

  /**
   * Subscribes `callback` to watch state updates; no operation if
   * `callback` is already subscribed to this state instance.
   * @param {function} callback It will be called without any arguments every
   * time the state content changes (note, howhever, separate state updates can
   * be applied to the state at once, and watching callbacks will be called once
   * after such bulk update).
   * @throws if {@link SsrContext} is attached to the state instance: the state
   * watching functionality is intended for client-side (non-SSR) only.
   */
  watch(callback) {
    if (this.ssrContext) throw new Error(ERR_NO_SSR_WATCH);

    const watchers = this.#watchers;
    if (watchers.indexOf(callback) < 0) {
      watchers.push(callback);
    }
  }
}
