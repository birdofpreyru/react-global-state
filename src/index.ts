import GlobalState from './GlobalState';

import GlobalStateProvider, {
  getGlobalState,
  getSsrContext,
} from './GlobalStateProvider';

import SsrContext from './SsrContext';

import useAsyncCollection, {
  type UseAsyncCollectionI,
  type UseAsyncCollectionResT,
} from './useAsyncCollection';

import {
  type AsyncDataEnvelopeT,
  type AsyncDataLoaderT,
  type AsyncDataReloaderT,
  type UseAsyncDataI,
  type UseAsyncDataOptionsT,
  type UseAsyncDataResT,
  newAsyncDataEnvelope,
  useAsyncData,
} from './useAsyncData';

import useGlobalState, { type UseGlobalStateI } from './useGlobalState';

export type { AsyncCollectionLoaderT, AsyncCollectionT } from './useAsyncCollection';

export type { SetterT, UseGlobalStateResT } from './useGlobalState';

export type { ForceT, ValueOrInitializerT } from './utils';

export {
  type AsyncDataEnvelopeT,
  type AsyncDataLoaderT,
  type AsyncDataReloaderT,
  type UseAsyncCollectionResT,
  type UseAsyncDataOptionsT,
  type UseAsyncDataResT,
  getGlobalState,
  getSsrContext,
  GlobalState,
  GlobalStateProvider,
  newAsyncDataEnvelope,
  SsrContext,
  useAsyncCollection,
  useAsyncData,
  useGlobalState,
};

interface API<
  StateT,
  SsrContextT extends SsrContext<StateT> = SsrContext<StateT>,
> {
  getGlobalState: typeof getGlobalState<StateT, SsrContextT>;
  getSsrContext: typeof getSsrContext<SsrContextT>,
  GlobalState: typeof GlobalState<StateT, SsrContextT>,
  GlobalStateProvider: typeof GlobalStateProvider<StateT, SsrContextT>,
  newAsyncDataEnvelope: typeof newAsyncDataEnvelope,
  SsrContext: typeof SsrContext<StateT>,
  useAsyncCollection: UseAsyncCollectionI<StateT>,
  useAsyncData: UseAsyncDataI<StateT>,
  useGlobalState: UseGlobalStateI<StateT>,
}

const api = {
  getGlobalState,
  getSsrContext,
  GlobalState,
  GlobalStateProvider,
  newAsyncDataEnvelope,
  SsrContext,
  useAsyncCollection,
  useAsyncData,
  useGlobalState,
};

export function withGlobalStateType<
  StateT,
  SsrContextT extends SsrContext<StateT> = SsrContext<StateT>,
>() {
  return api as API<StateT, SsrContextT>;
}
