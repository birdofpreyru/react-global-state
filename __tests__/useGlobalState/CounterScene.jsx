/**
 * Tests `useGlobalState(..)` functionality in the "CounterScene":
 * one component shows a value from the global state, another component
 * implements a button, which updates that state when clicked.
 */
/* eslint-disable react/prop-types */

import pretty from 'pretty';

import GlobalState from 'GlobalState';
import { act, mount, unmount } from 'jest/utils';
import { GlobalStateProvider, useGlobalState } from 'src';

jest.useFakeTimers();

function CounterView() {
  const [count] = useGlobalState('counter', 0);
  return <div>{count}</div>;
}

function CounterButton({ testButtonId }) {
  const [count, setCount] = useGlobalState('counter', 0);
  return (
    <div>
      <button
        data-testid={testButtonId}
        onClick={() => setCount(1 + count)}
        type="button"
      >
        Bump!
      </button>
    </div>
  );
}

/**
 * This is the test scene.
 * @param {Boolean} [props.innerProviderIsProxy] Makes the inner provider proxy.
 */
function TestScene({ stateProxy }) {
  return (
    <GlobalStateProvider>
      <CounterView />
      <CounterButton testButtonId="button-1" />
      <GlobalStateProvider stateProxy={stateProxy}>
        <CounterView />
        <CounterButton testButtonId="button-2" />
      </GlobalStateProvider>
    </GlobalStateProvider>
  );
}

let scene = null;
afterEach(() => {
  if (scene) {
    unmount(scene);
    scene = null;
  }
});

test('Throws if GlobalStateProvider is missing', () => {
  const consoleError = console.error;
  console.error = () => null;
  let message;
  try {
    mount(<CounterView />);
  } catch (error) {
    ({ message } = error);
  }
  console.error = consoleError;
  expect(message).toMatchSnapshot();
});

test('The scene works as expected', () => {
  act(() => {
    scene = mount(<TestScene />);
  });
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
  let button = document.querySelector('[data-testid=button-1]');
  act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    jest.runAllTimers();
  });
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
  button = document.querySelector('[data-testid=button-2]');
  act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    jest.runAllTimers();
  });
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
});

test('Test of `stateProxy` prop of the state provider', () => {
  act(() => {
    scene = mount(<TestScene stateProxy />);
  });
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
  let button = document.querySelector('[data-testid=button-1]');
  act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    jest.runAllTimers();
  });
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
  button = document.querySelector('[data-testid=button-2]');
  act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    jest.runAllTimers();
  });
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
});

test('Test of `stateProxy` prop of the state provider II', () => {
  const innerGlobalState = new GlobalState();
  act(() => {
    scene = mount(<TestScene stateProxy={innerGlobalState} />);
  });
  expect(innerGlobalState).toMatchSnapshot();
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
  let button = document.querySelector('[data-testid=button-1]');
  act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    jest.runAllTimers();
  });
  expect(innerGlobalState).toMatchSnapshot();
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
  button = document.querySelector('[data-testid=button-2]');
  act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    jest.runAllTimers();
  });
  expect(innerGlobalState).toMatchSnapshot();
  expect(pretty(scene.innerHTML)).toMatchSnapshot();
});
