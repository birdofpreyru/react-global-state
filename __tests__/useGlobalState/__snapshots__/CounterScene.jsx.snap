// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of \`stateProxy\` prop of the state provider 1`] = `
"<div>0</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>0</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`Test of \`stateProxy\` prop of the state provider 2`] = `
"<div>1</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>1</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`Test of \`stateProxy\` prop of the state provider 3`] = `
"<div>2</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>2</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`Test of \`stateProxy\` prop of the state provider II 1`] = `
Object {
  "counter": 0,
}
`;

exports[`Test of \`stateProxy\` prop of the state provider II 2`] = `
"<div>0</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>0</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`Test of \`stateProxy\` prop of the state provider II 3`] = `
Object {
  "counter": 0,
}
`;

exports[`Test of \`stateProxy\` prop of the state provider II 4`] = `
"<div>1</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>0</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`Test of \`stateProxy\` prop of the state provider II 5`] = `
Object {
  "counter": 1,
}
`;

exports[`Test of \`stateProxy\` prop of the state provider II 6`] = `
"<div>1</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>1</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`The scene works as expected 1`] = `
"<div>0</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>0</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`The scene works as expected 2`] = `
"<div>1</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>0</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`The scene works as expected 3`] = `
"<div>1</div>
<div><button data-testid=\\"button-1\\" type=\\"button\\">Bump!</button></div>
<div>1</div>
<div><button data-testid=\\"button-2\\" type=\\"button\\">Bump!</button></div>"
`;

exports[`Throws if GlobalStateProvider is missing 1`] = `"Missing GlobalStateProvider"`;
